orttraining/orttraining/core/session/training_session.cc

```diff
@@ -880,6 +891,59 @@ common::Status TrainingSession::GetStateTensors(NameMLValMap& state_tensors) {
   return GetSessionState().GetInitializedTensors(GetStateTensorNames(), allow_missing, state_tensors);
 }
 
+common::Status TrainingSession::GetOptimizerState(std::unordered_map<std::string, NameMLValMap>& opt_state_tensors) {
+  bool allow_missing = (opt_graph_config_.deepspeed_zero.stage != 0);
+  for (const auto& weight_map : weight_to_opt_mapping_) {
+    NameMLValMap curr_opt_tensors;
+    std::unordered_set<std::string> opt_names(weight_map.second.begin(), weight_map.second.end());
+    const auto& weight_name = weight_map.first;
+    GetSessionState().GetInitializedTensors(opt_names, allow_missing, curr_opt_tensors);
+    opt_state_tensors[weight_name] = {};
+    // Keep only prefix in returned value
+    for (const auto& opt_pair: curr_opt_tensors) {
+      const auto& opt_name = opt_pair.first;
+      std::string pre_fix_only = opt_name.substr(0, opt_name.find(weight_name) - 1);
+      opt_state_tensors[weight_name][pre_fix_only] = opt_pair.second;
+    }
+  }
+  return Status::OK();
+}
+
+common::Status TrainingSession::GetModelState(std::unordered_map<std::string, NameMLValMap>& model_state_tensors, bool include_mixed_precision_weights) {
+  bool allow_missing = (opt_graph_config_.deepspeed_zero.stage != 0);
+  std::unordered_set<std::string> fp_tensor_names{};
+  fp_tensor_names.insert(
+      weights_to_train_.begin(), weights_to_train_.end());
+  NameMLValMap fp_weights;
+  GetSessionState().GetInitializedTensors(fp_tensor_names, allow_missing, fp_weights);
+  model_state_tensors.emplace("full_precision", fp_weights);
+  if (include_mixed_precision_weights) {
+    std::unordered_set<std::string> mp_tensor_names{};
+    mp_tensor_names.insert(
+        mixed_precision_weight_initializer_names_.begin(), mixed_precision_weight_initializer_names_.end());
+    NameMLValMap mp_weights;
+    GetSessionState().GetInitializedTensors(mp_tensor_names, allow_missing, mp_weights);
+    model_state_tensors.emplace("mixed_precision", mp_weights);
+  }
+  return Status::OK();
+}
+
+// use dimension of fp32 weights
+common::Status TrainingSession::GetPartitionInfoMap(std::unordered_map<std::string, std::unordered_map<std::string, std::vector<int>>>& part_info_map) {
+  bool allow_missing = (opt_graph_config_.deepspeed_zero.stage != 0);
+  std::unordered_set<std::string> fp_tensor_names{};
+  fp_tensor_names.insert(
+      weights_to_train_.begin(), weights_to_train_.end());
+  NameMLValMap fp_weights;
+  GetSessionState().GetInitializedTensors(fp_tensor_names, allow_missing, fp_weights);
```

Comment:
Why do we need to fetch the initializer values? if we just need the weight names, we already have them in weights_to_train_, which we also do not need as we should populate this only for the weights that have been partitioned. Another thing that we have to return according to the design is the original shapes of the partitioned tensor. I would recommend maintaining a PartitionInfo struct, passing it to both the Megatron transformer and the Zero optimizer graph builder and populating the struct in those transforms. You can use that to just populate the output in this method.

Reply:
Somehow I thought the dimensions of fp32 weights will give the correct dimension. Anyway this is changed now